src/main/java/org/topbraid/shacl/tools/Infer.java

```diff
@@ -43,15 +43,51 @@ public static void main(String[] args) throws IOException {
 		System.setErr(oldPS);
 		infer.run(args);
 	}
-	
-	
+
+
 	private void run(String[] args) throws IOException {
 		Model dataModel = getDataModel(args);
 		Model shapesModel = getShapesModel(args);
 		if(shapesModel == null) {
 			shapesModel = dataModel;
 		}
-		Model results = RuleUtil.executeRules(dataModel, shapesModel, null, null);
-		results.write(System.out, FileUtils.langTurtle);
-	}
+        int maxIterations = getMaxIterations(args);
+
+        // execute the rules over and over until there are no new results
+        // or until a maximum number of iterations has been reached.
+        // Keep track of all results from each iteration in a separate model.
+        // Break if no new results are found, or if the size of the allResults
+        // model does not increase after a run.
+
+        // stores which iteration we are on
+        int iteration = 0;
+        // stores all results from each iteration
+        Model results = null;
+        do {
+            // execute the rules
+            Model newResults = RuleUtil.executeRules(dataModel, shapesModel, null, null);
+            // if this is the first iteration, set the results model, otherwise add to it
+            if (results == null) {
+                results = newResults;
+            } else {
+                results.add(newResults);
+            }
+            // if no new results were found, break
+            if (newResults.size() == 0) {
+                break;
+            }
+            // if the size of the allResults model did not increase, break
+            // (this means that the new results were already in the allResults model)
+            if (results.size() == newResults.size()) {
+                break;
+            }
+            // otherwise, continue by adding the new results to the data model, incrementing the iteration,
+            // and continuing the loop to see if more inferences are performed
+            dataModel.add(newResults);
```

Comment:
This .add only needs to happen if there is another iteration. Yet another alternative is to add a memory graph to the subgraphs of the OntModel and pass it in as third argument to the RuleUtil.executeRules method.

Reply:
Good point! I implemented the first fix you mentioned, but I can switch it to the memory graph if you'd prefer